### DISCUSSION TITLE
How to Apply Multiple Colors to a Waveform Based on Speaker Start and End Times?

### DISCUSSION THREAD

**Discussion by jawwad-dev** [katspaugh/wavesurfer.js#3839] [Q&A] [answered]
Are there any options available to apply multiple colors to a waveform based on specific start and end times? I have an audio file with multiple speakers, and I know the exact times when each speaker begins and ends. I want to visually differentiate each speaker by using different colors in the waveform. Is there a way to achieve this?

**Comment by katspaugh** [comment, marked as answer]
You can use the Regions plugin with regions’ color set to the background color + add a mix-blend-mode in CSS.

Or write a custom render function. See https://wavesurfer.xyz/examples/?custom-render.js

**Reply by jawwad-dev** [reply]
Thank you

**Reply by jawwad-dev** [reply]
How can I ensure that colors applied to my waveform based on time stamps remain consistent when zooming in? Currently, when I zoom, the colors are being rendered based on width instead of the specified time stamps. What adjustments can I make to my render function to maintain accurate color mapping regardless of the zoom level?

To give multiple colors to waveform I'm using renderFunction. 
cc: @katspaugh 

https://github.com/user-attachments/assets/1c8f874c-34d1-4158-8bbd-6eae87af6110

**Reply by katspaugh** [reply]
When zoomed, there's currently no way to tell where in the waveform you are (except for a hacky one, see further).

We should add an offset parameter to the render function [here](https://github.com/katspaugh/wavesurfer.js/blob/main/src/renderer.ts#L497).

A hacky workaround would be to read the left offset of the current canvas like this:

```js
const offset = ctx.canvas.offsetLeft
const totalWidth = wavesurfer.getWrapper().clientWidth
const startTime = (offset / totalWidth) * wavesurfer.getDuration()
```

**Reply by jawwad-dev** [reply]
For now may main concern is with the colors with respect to time stamps. Here I'm calculating the width on the basis of time and initially its working fine. But after zoom in color width not increasing on the basis of time it's remain same. 

Also as you mentioned above offset, it's not clear for me how can we use offset for zoom. 


```import WaveSurfer from 'wavesurfer.js';
import Zoom from 'wavesurfer.js/dist/plugins/zoom.esm.js';

const speakerTimes = [
  { start: 0, end: 6, color: 'rgba(255, 0, 0, 0.3)' },  
  { start: 6, end: 15, color: 'rgba(0, 255, 0, 0.3)' }, 
  { start: 15, end: 26, color: 'rgba(0, 0, 255, 0.3)' }
];

const wavesurfer = WaveSurfer.create({
  container: document.body,
  waveColor: 'rgb(200, 0, 200)',  // Default wave color
  progressColor: 'rgb(100, 0, 100)',
  url: '/examples/audio/demo.wav',
  plugins: [Zoom.create({})],


  renderFunction: (channels, ctx) => {
    let vScale = 1;
    const max = Array.from(channels[0]).reduce(
      (max, value) => Math.max(max, Math.abs(value)),
      0
    );
    vScale = max ? 1 / max : 1;

    const topChannel = channels[0];
    const bottomChannel = channels[1] || channels[0];
    const length = topChannel.length;
    const { width, height } = ctx.canvas;
    const halfHeight = height / 2;

    const barWidth = 4;
    const barGap = 2;
    const barRadius = 2;

    const barIndexScale = width / (barWidth + barGap) / length;
    const rectFn = barRadius && 'roundRect' in ctx ? 'roundRect' : 'rect';

    // Calculate the total duration of the audio
    const totalDuration = wavesurfer.getDuration();

    let prevX = 0;
    let maxTop = 0;
    let maxBottom = 0;

    for (let i = 0; i <= length; i++) {
      const x = Math.round(i * barIndexScale);
      if (x > prevX) {
        const topBarHeight = Math.round(maxTop * halfHeight * vScale);
        const bottomBarHeight = Math.round(
          maxBottom * halfHeight * vScale
        );
        const barHeight = topBarHeight + bottomBarHeight || 1;
        // Vertical alignment
        let y = halfHeight - topBarHeight;

        // Calculate the corresponding time for the current bar
        const time = (i / length) * totalDuration;

        // Check if the current time is within any of the speaker times
        const speakerTime = speakerTimes.find(
          ({ start, end }) => time >= start && time <= end
        );

        // Set the fillStyle based on the speaker's time or default wave color
        ctx.fillStyle = speakerTime ? speakerTime.color : 'rgb(200, 0, 200)';

        ctx.beginPath();
        ctx[rectFn](
          prevX * (barWidth + barGap),
          y,
          barWidth,
          barHeight,
          barRadius
        );
        ctx.fill();
        ctx.closePath();

        prevX = x;
        maxTop = 0;
        maxBottom = 0;
      }
      const magnitudeTop = Math.abs(topChannel[i] || 0);
      const magnitudeBottom = Math.abs(bottomChannel[i] || 0);
      if (magnitudeTop > maxTop) maxTop = magnitudeTop;
      if (magnitudeBottom > maxBottom) maxBottom = magnitudeBottom;
    }
  }
});

wavesurfer.on('interaction', () => {
  wavesurfer.play();
});
```

cc: @katspaugh

**Reply by jawwad-dev** [reply]
Initially we have only three colors with respect to time stamps. When I zoom in it's start adding colors again from right side. I want to zoom in and from 0-6 seconds color must remain same like red. From 6-15 seconds it remain same as green and so on. 

https://github.com/user-attachments/assets/dc9f7006-b5e7-47e5-aa62-8defa43333ab
